src/Rules/FunctionDefinitionCheck.php

```diff
@@ -334,6 +334,17 @@ private function checkParametersAcceptor(
 
 					return $traverse($type);
 				});
+
+				if ($parameter instanceof ParameterReflectionWithPhpDocs && $parameter->getOutType() !== null) {
+					TypeTraverser::map($parameter->getOutType(), static function (Type $type, callable $traverse) use (&$templateTypes): Type {
+						if ($type instanceof TemplateType) {
+							unset($templateTypes[$type->getName()]);
```

Comment:
This change wasn't needed for `@phpstan-assert` and doesn't make sense for `@param-out` either.

With a signature like this:

```php
/**
 * @template S of self
 * @param-out S $s
 */
function genericSelf(mixed &$s): void
```

The generics here don't do anything because it's the same as just writing `@param-out mixed $s`.

Reply:
not sure I get your point. we need to unset the `$templateTypes` here, as we otherwise get a error below

> Template type S of function ParamOutTemplate\genericSelf() is not referenced in a parameter.

Reply:
Yes, which is expected, because the generics for this function are useless. S will always be self.